executor/src/witgen/range_constraints.rs

```diff
@@ -185,6 +190,60 @@ fn range_multiple<T: FieldElement>(min: T, max: T, factor: T) -> (T, T) {
     }
 }
 
+/// Computes the intersection of two intervals.
+/// There are cases where the intersection cannot be represented as a single internal.
+/// in that case, it returns the smaller of the two inputs.
+/// If the intersection is empty, returns None.
+fn interval_intersection<T: FieldElement>(mut a: (T, T), mut b: (T, T)) -> Option<(T, T)> {
+    // We shift both intervals until they are both non-wrapping intervals.
+    // If we do not succeed after shifting both of them by the smallest amount,
+    // it means that the intersection cannot be expressed as a single interval.
+    // In that case we just choose the smaller of the two inputs.
+    let shift: T = if a.0 <= a.1 {
+        // a is regular
+        if b.0 <= b.1 {
+            0.into()
+        } else if a.1 <= b.1 || b.1 <= a.0 {
+            b.1
+        } else if a.1 < b.0 {
+            a.1
+        } else {
+            // intersection consists of two intervals.
+            if range_width(a.0, a.1) >= range_width(b.0, b.1) {
+                return Some(a);
+            } else {
+                return Some(b);
+            }
+        }
+    } else if b.0 <= b.1 {
+        // a is wrapping b is regular
+        return interval_intersection(b, a);
+    } else {
+        // both a and b are wrapping
+        if cmp::max(a.1, b.1) < cmp::min(a.0, b.0) {
+            cmp::max(a.1, b.1)
```

Comment:
Put this in a variable? I understand it might be more legible like this, so it is a matter of preference...